File: ios/Talise/Features/Ramps/RoundedFlag.swift
import SwiftUI

extension View {
    /// The standard ramp card chrome — soft surface, radius-20, hairline ring.
    /// Shared across the Ramps module (on-ramp + cash-out screens).
    func rampCard() -> some View {
        self
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: 20, style: .continuous)
                    .fill(TaliseColor.surface)
            )
            .overlay(
                RoundedRectangle(cornerRadius: 20, style: .continuous)
                    .strokeBorder(TaliseColor.line, lineWidth: 1)
            )
    }
}

/// A flag emoji rendered as a clean, rounded chip — a circular disc with the
/// flag centered and a hairline ring. This is the app's standard "rounded
/// flag" used across the ramps (and reusable anywhere a country needs a
/// compact, on-brand glyph). Emoji flags don't clip nicely on their own, so
/// we center the glyph on a soft surface disc instead of trying to mask it.
struct RoundedFlag: View {
    let flag: String
    var size: CGFloat = 40
    /// Dim the chip (used for unavailable / "soon" corridors).
    var dimmed: Bool = false

    var body: some View {
        Text(flag)
            .font(.system(size: size * 0.58))
            .frame(width: size, height: size)
            .background(Circle().fill(TaliseColor.surface2))
            .overlay(Circle().strokeBorder(TaliseColor.line, lineWidth: 1))
            .clipShape(Circle())
            .saturation(dimmed ? 0.15 : 1)
            .opacity(dimmed ? 0.55 : 1)
    }
}

/// A small cluster of overlapped rounded flags — the compact "coming soon"
/// treatment so a long tail of not-yet-live corridors reads as one quiet row
/// of country circles rather than a wall of disabled list items.
struct OverlappedFlags: View {
    let flags: [String]
    var size: CGFloat = 32
    var max: Int = 5

    var body: some View {
        HStack(spacing: -size * 0.32) {
            ForEach(Array(flags.prefix(max).enumerated()), id: \.offset) { _, f in
                Text(f)
                    .font(.system(size: size * 0.56))
                    .frame(width: size, height: size)
                    .background(Circle().fill(TaliseColor.surface2))
                    .overlay(Circle().strokeBorder(TaliseColor.bg, lineWidth: 2))
                    .clipShape(Circle())
                    .saturation(0.15)
                    .opacity(0.55)
            }
            if flags.count > max {
                Text("+\(flags.count - max)")
                    .font(TaliseFont.mono(11, weight: .medium))
                    .foregroundStyle(TaliseColor.fgDim)
                    .frame(width: size, height: size)
                    .background(Circle().fill(TaliseColor.surface2))
                    .overlay(Circle().strokeBorder(TaliseColor.bg, lineWidth: 2))
                    .clipShape(Circle())
                    .padding(.leading, size * 0.32 + 4)
            }
        }
    }
}

/// One selectable corridor row: rounded flag + name + currency/rail subtitle,
/// a trailing chevron, and a clean tappable card. Disabled rows aren't built
/// here — "soon" corridors collapse into `OverlappedFlags` instead.
struct CorridorRow: View {
    let corridor: RampCorridor
    var selected: Bool = false

    var body: some View {
        HStack(spacing: 14) {
            RoundedFlag(flag: corridor.flag, size: 40)
            VStack(alignment: .leading, spacing: 2.5) {
                HStack(spacing: 7) {
                    Text(corridor.name)
                        .font(TaliseFont.heading(16, weight: .semibold))
                        .kerning(-0.3)
                        .foregroundStyle(TaliseColor.fg)
                    Text(corridor.currencyCode)
                        .font(TaliseFont.mono(10, weight: .regular))
                        .kerning(0.6)
                        .foregroundStyle(TaliseColor.fgDim)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(Capsule().fill(TaliseColor.surface2))
                }
                Text(corridor.railLabel)
                    .font(TaliseFont.body(12.5, weight: .light))
                    .foregroundStyle(TaliseColor.fgMuted)
            }
            Spacer(minLength: 0)
            if selected {
                Image(systemName: "checkmark")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(TaliseColor.greenMint)
            } else {
                Image(systemName: "chevron.right")
                    .font(.system(size: 11, weight: .medium))
                    .foregroundStyle(TaliseColor.fgDim)
            }
        }
        .padding(.horizontal, 18)
        .padding(.vertical, 14)
        .background(
            RoundedRectangle(cornerRadius: 20, style: .continuous)
                .fill(TaliseColor.surface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 20, style: .continuous)
                .strokeBorder(
                    selected ? TaliseColor.greenMint.opacity(0.5) : TaliseColor.line,
                    lineWidth: 1
                )
        )
        .contentShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
    }
}

File: android/app/src/main/java/io/talise/app/feature/ramps/RoundedFlag.kt
package io.talise.app.feature.ramps

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.graphics.ColorMatrix
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.talise.app.R
import io.talise.app.ui.components.rampCard
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType

/**
 * Rounded circular country flags for the ramps, ported 1:1 from iOS
 * `RoundedFlag.swift` (same vendored circle-flags set the web app uses,
 * bundled as `res/drawable-nodpi/flag_<cc>.png`).
 */

/** Drawable for an ISO alpha-2 code (or "EU"); null when the asset is missing. */
internal fun flagRes(code: String): Int? = when (code.lowercase()) {
    "ae" -> R.drawable.flag_ae
    "bd" -> R.drawable.flag_bd
    "ca" -> R.drawable.flag_ca
    "de" -> R.drawable.flag_de
    "dz" -> R.drawable.flag_dz
    "eg" -> R.drawable.flag_eg
    "eu" -> R.drawable.flag_eu
    "fr" -> R.drawable.flag_fr
    "gb" -> R.drawable.flag_gb
    "gh" -> R.drawable.flag_gh
    "id" -> R.drawable.flag_id
    "in" -> R.drawable.flag_in
    "jp" -> R.drawable.flag_jp
    "ke" -> R.drawable.flag_ke
    "ma" -> R.drawable.flag_ma
    "ng" -> R.drawable.flag_ng
    "ph" -> R.drawable.flag_ph
    "pk" -> R.drawable.flag_pk
    "sa" -> R.drawable.flag_sa
    "sg" -> R.drawable.flag_sg
    "us" -> R.drawable.flag_us
    "vn" -> R.drawable.flag_vn
    "za" -> R.drawable.flag_za
    else -> null
}

private fun desaturation(saturation: Float): ColorFilter =
    ColorFilter.colorMatrix(ColorMatrix().apply { setToSaturation(saturation) })

/**
 * A circular country flag, the app's standard "rounded flag" across the ramps.
 * `code` is an ISO alpha-2 (or "EU"); a missing asset falls back to a neutral
 * disc so nothing breaks. Mirrors iOS `RoundedFlag`.
 */
@Composable
fun RoundedFlag(
    code: String,
    size: Dp = 40.dp,
    /** Dim the chip (used for unavailable / "soon" corridors). */
    dimmed: Boolean = false,
) {
    val res = flagRes(code)
    Box(
        Modifier
            .size(size)
            .clip(CircleShape)
            .background(TaliseColors.surface2, CircleShape) // fallback disc
            .border(1.dp, TaliseColors.line, CircleShape)
            .alpha(if (dimmed) 0.6f else 1f),
        contentAlignment = Alignment.Center,
    ) {
        if (res != null) {
            Image(
                painterResource(res),
                contentDescription = null,
                modifier = Modifier.size(size).clip(CircleShape),
                contentScale = ContentScale.Crop,
                colorFilter = if (dimmed) desaturation(0.2f) else null,
            )
        }
    }
}

/**
 * A small cluster of overlapped circular flags, the compact "coming soon"
 * treatment so a long tail of not-yet-live corridors reads as one quiet row
 * of country circles rather than a wall of disabled list items. `codes` are
 * ISO alpha-2 codes. Mirrors iOS `OverlappedFlags`.
 */
@Composable
fun OverlappedFlags(
    codes: List<String>,
    size: Dp = 32.dp,
    max: Int = 6,
) {
    Row {
        codes.take(max).forEachIndexed { index, cc ->
            Box(
                Modifier
                    .offset(x = -(size * 0.34f) * index)
                    .size(size)
                    .clip(CircleShape)
                    .background(TaliseColors.surface2, CircleShape)
                    .border(2.dp, TaliseColors.bg, CircleShape)
                    .alpha(0.7f),
                contentAlignment = Alignment.Center,
            ) {
                flagRes(cc)?.let { res ->
                    Image(
                        painterResource(res),
                        contentDescription = null,
                        modifier = Modifier.size(size).clip(CircleShape),
                        contentScale = ContentScale.Crop,
                        colorFilter = desaturation(0.25f),
                    )
                }
            }
        }
        if (codes.size > max) {
            Box(
                Modifier
                    .offset(x = -(size * 0.34f) * (max - 1) + 4.dp)
                    .size(size)
                    .clip(CircleShape)
                    .background(TaliseColors.surface2, CircleShape)
                    .border(2.dp, TaliseColors.bg, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    "+${codes.size - max}",
                    style = TaliseType.mono(11.sp, FontWeight.Medium),
                    color = TaliseColors.fgDim,
                )
            }
        }
    }
}

/**
 * One selectable corridor row: rounded flag + name + currency/rail subtitle,
 * a trailing chevron, and a clean tappable card. Disabled rows aren't built
 * here, "soon" corridors collapse into [OverlappedFlags] instead. Mirrors iOS
 * `CorridorRow`.
 */
@Composable
fun CorridorRow(
    corridor: RampCorridor,
    selected: Boolean = false,
) {
    val ring = if (selected) TaliseColors.greenMint.copy(alpha = 0.5f) else TaliseColors.line
    Row(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(20.dp))
            .background(TaliseColors.surface, RoundedCornerShape(20.dp))
            .border(1.dp, ring, RoundedCornerShape(20.dp))
            .padding(horizontal = 18.dp, vertical = 14.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        RoundedFlag(code = corridor.code, size = 40.dp)
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.5.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(7.dp)) {
                Text(
                    corridor.name,
                    style = TaliseType.heading(16.sp, FontWeight.SemiBold),
                    letterSpacing = (-0.3).sp,
                    color = TaliseColors.fg,
                )
                Text(
                    corridor.currencyCode,
                    style = TaliseType.mono(10.sp),
                    letterSpacing = 0.6.sp,
                    color = TaliseColors.fgDim,
                    modifier = Modifier
                        .background(TaliseColors.surface2, CircleShape)
                        .padding(horizontal = 6.dp, vertical = 2.dp),
                )
            }
            Text(
                corridor.railLabel,
                style = TaliseType.body(12.5.sp, FontWeight.Light),
                color = TaliseColors.fgMuted,
            )
        }
        if (selected) {
            Icon(
                Icons.Filled.Check,
                contentDescription = null,
                tint = TaliseColors.greenMint,
                modifier = Modifier.size(13.dp),
            )
        } else {
            Icon(
                Icons.AutoMirrored.Filled.KeyboardArrowRight,
                contentDescription = null,
                tint = TaliseColors.fgDim,
                modifier = Modifier.size(16.dp),
            )
        }
    }
}

/** iOS `messageCard(title:body:)` clone shared by the ramp flows. */
@Composable
internal fun RampMessageCard(title: String, body: String) {
    Column(
        Modifier.fillMaxWidth().rampCard().padding(18.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(title, style = TaliseType.heading(16.sp, FontWeight.SemiBold), color = TaliseColors.fg)
        Text(body, style = TaliseType.body(13.sp, FontWeight.Light), color = TaliseColors.fgMuted)
    }
}
